[24123] Work package group form configuration

machisuji opened the pull request:
https://community.openproject.com/projects/openproject/work_packages/24123
https://community.openproject.com/projects/openproject/work_packages/24204

machisuji committed:
attribute groups backend draft 1

machisuji committed:
fixed _attributeGroups format

machisuji committed:
don't make default attr groups dependent on attr visibility

wielinde committed:
wip: angularjs-dragula for form_configuration

wielinde committed:
the groups have an order

wielinde committed:
wip: form_configuration

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/1dc19b658bc521e6865c4f42412c6d8b6f6bfd6d) (deployed commit: cebbc6768cfe44930093fa75c79d36b932944279).

wielinde committed:
More mocking of the UI

wielinde committed:
Change Migration: Move from Hash to Array

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/4f17430a245af74e66f910a237239f033aef4276) (deployed commit: 1b01bdfba0351a0ee4150a6d7c44db3325e77fee).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/672008c9a2b1e009af3bc2482f344a9471387202) (deployed commit: 1b01bdfba0351a0ee4150a6d7c44db3325e77fee).

machisuji committed:
values must be an array

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/b2e3710e79ff753b1f0c2c85dd832bc7cd52413c) (deployed commit: 0409e2d83e6c083a5eb1087f45aa5cb57a31345d).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/69c18a555c699ce0296e508dd43c3b26363c9f9c) (deployed commit: 0409e2d83e6c083a5eb1087f45aa5cb57a31345d).

wielinde committed:
Don't carry empty attribute groups around.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/f752304d58c6ba62ab9bb6edd637da92b5164e1a) (deployed commit: 811ccfbedee7986c7291c0f2da57c2c10fe43115).

wielinde committed:
Adding some specs for #default_attribute_groups

codeclimate

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/9a2d766f03d497a93ddfe7f31e9f07c8c40f9871) (deployed commit: a59ed981e7da939a34c3eaafe7d2d8c59922e157).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/496bac2197341127144fc90e9bb6883a82374141) (deployed commit: 8649310af9bc0621694f85f0e28364a6461d3a81).

wielinde committed:
fixing specs for new format of saving attribute_groups as an array of arrays

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/0606dcbe0c9c64ea39178f74a0d3e5fef739f06d) (deployed commit: 17554360cbc8eda40fbbb47043d95826c7da9aae).

wielinde committed:
move dragula logic to angular controller

wielinde committed:
inactive attribute group plus translations

wielinde committed:
codeclimate

wielinde committed:
Dom structure to input type hidden

wielinde committed:
Frontend logic for addGroup, deleteGroup, deactivateAttribute, and reset to defaults.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/6a7546bcfd834b852696fa621223f592012c5311) (deployed commit: 9daa4abc387de712aff06e78167ad6c884056f15).

wielinde committed:
WIP default_atttibute_groups filtered

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/1ac925802119a8efa27dccb53c7cfdc91a62f45d) (deployed commit: 52bcb62b93d89088ae74a12914022193a159fa48).

wielinde committed:
Not necessarry stuff killed

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/18027e2e4165219c15c15658d3e1664cc6460ca1) (deployed commit: 933a2f8a191decb53001b786e5e72a9f1b04261a).

wielinde committed:
Latest WIP [ci skip]

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/99b78d2bc51a70863dddd6f7092d676cd5d052f9) (deployed commit: f2be13f082ab013a438773fccffb8125796640df).

wielinde committed:
Fix migration [ci skip]

wielinde committed:
Fix: typo in default_attribute_groups

wielinde committed:
remove debugging stuff [ci skip]

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/f411a916fcc93e1cb28be754feb3bfbac13e73c2) (deployed commit: 9f1282ad890f1749578f56ebc601b409c11ad1f9).

wielinde committed:
Add validation for field `attribute_groups`

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/38d0f30faf3e96440c8cd6c250e39938b580bf87) (deployed commit: 930680ba5f47a355253e9bb47c4c060b5d84e7cf).

wielinde committed:
Fix types_controller_spec.rb

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/369c68d502647e07cbe07b4a31d05f917b921a89) (deployed commit: 21a498f93ee386b7656c355e0d2ade77c94f357d).

wielinde committed:
Finished edit in place

wielinde committed:
fix legacy spec

wielinde committed:
Fix redirects after creating a type.

wielinde committed:
Fix cukes for types

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/8bea0d74dacaa529a9e18e4b95bbcda3315be4dd) (deployed commit: 8701105273803c85c190d5ede2a72143cb35a052).

wielinde committed:
Buttons; No empty groups; UI deletion of groups; Name new group;

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/172b438e326a25d4477a2137ac90b17d4446d2d4) (deployed commit: c78ffe5e0f5e2b135c1313b336a8721700c34270).

wielinde committed:
Fixing some specs

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/4de77fb425aaa6cff96e3e6e3cd48f6103f3e779) (deployed commit: 6d0e50a358cbf5afdeb58762b7b157636c024b43).

wielinde committed:
Fixing spec

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/459c611d50cc2cc884f159a8a908c1484ade695c) (deployed commit: 1261d375bc1cdee71dceea5b017fc2784a057329).

oliverguenther committed:
Implement frontend for group configuration

oliverguenther committed:
Fix single view date references

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/8f2acd097d78c237db60877d7760374e12b1d359) (deployed commit: b5707d470545ee78faa1586e186ae9b885c27964).

wielinde committed:
Typo in file name

wielinde committed:
Hopefully fixing a bug with renaming attribute groups with spaces in the name

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/443d2e84e7ea5d66e979fec837de6de0bd26f23f) (deployed commit: be61bb654c71b1f7727fad3a3a0a227b5a84690c).

oliverguenther committed:
Place required fields in separate, fixed group above description

oliverguenther committed:
Move type to subject row

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/d5132ce16b59ed6f6471bc5f012fc051a72f559a) (deployed commit: 2ec47db23092d919d412b6ae77452787e37719e3).

oliverguenther committed:
Add some minor styling

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/9ce656ff0a784808beda1da876828950d2348929) (deployed commit: 1859071a1d2f3bc63639703fef7ba1f43dd72e77).

wielinde committed:
First few feature specs for form_configuration

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/902820b16e6df5464bc4e78d153675235cee2f6b) (deployed commit: 66e855cfd5f837b2a0303df95850cf072fd38df6).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/9e295b2bd7addbbb96ece969d6adac780abc627d) (deployed commit: 66e855cfd5f837b2a0303df95850cf072fd38df6).

oliverguenther committed:
Hide required fields with default correctly

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/5394e67545612014af70463c20314ae7986b7506) (deployed commit: 2d9578d90f1a2bb5e1c969225db034310b39aee9).

oliverguenther committed:
set flex-basis on type in subject row

oliverguenther committed:
Move panel-expander upwards

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/483d6627a72f84330ff58bb362090ed3362ac692) (deployed commit: 92f54be0f9c0f75318707943be2217c68aab8572).

oliverguenther committed:
Remove flex-basis

Too much visible space

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/433681ee9035949e42fa322d3520e7c6b8676ecf) (deployed commit: 9c8c0d3f3870412d5146bfe81f5dc7bf95533d86).

oliverguenther committed:
Add custom sub-models for attribute constraints and groups

This allows plugins such as backlogs and costs to integrate
into the grouping while restricting while they are visible.

The problem here is that on a global admin page, the plugin's attributes
should be available, however in a project context, they should only
be added when the respective module is active.

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/0a0af9add6c8ff13ca65fcb74646a2e94688e9f2) (deployed commit: 72d930cd7a149321457e35dcb9bffc0b35a442dc).

oliverguenther committed:
Correct translations for default groups

oliverguenther committed:
Remove WP Overview service

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/b678225b81d908e230413d5537785a5defdf0a9e) (deployed commit: 6fb0a3fb384099ba41cbc92d8503ae502bb940e7).

oliverguenther committed:
Fix AR method name duplication

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/6533de19cfdb1cfa69b523d06b195fbc689721de) (deployed commit: 14515a0b6bda9e4294f1ef85baba1bc8910066da).

oliverguenther committed:
Fix tests

oliverguenther committed:
Remove unused helper

oliverguenther committed:
Only hide group if all members SHOULD be hidden

oliverguenther committed:
Fix more tests

oliverguenther committed:
Extend form configuration spec

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/5bb0f8da30bf701a5e9870e13ee3c9d36d10a8f4) (deployed commit: 5211e8bcdcabb6778088d2aff27ab46da916c9e9).

oliverguenther commented:
@wielinde Could you take a look at this PR?

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/595cc216e24c4042554fadc6294f8925f518627d) (deployed commit: 5211e8bcdcabb6778088d2aff27ab46da916c9e9).

wielinde commented on frontend/app/components/work-packages/wp-single-view/wp-single-view.directive.html:
```diff
@@ -23,81 +52,64 @@
-                        expand-text="{{ $ctrl.I18n.t('js.label_show_attributes') }}"
-                        collapse-text="{{ $ctrl.I18n.t('js.label_hide_attributes') }}">
-        </panel-expander>
+            ng-bind="group.name"></h3>
```
@oliverguenther Here we have an issue with the default groups. For instance it show "ESTIMATES_AND_TIME" instead of "ESTIMATES AND TIME" (note the spaces instead of the underscores. In the the template `_form_configuration.html.erb` I therefore added another data structure that allows for translations. I am not sure what the best solution would be. Maybe the translation needs to be part of the schema.

oliverguenther replied:
The groups will always contain the translated name. There is no mapping from an internal group key (e.g., `:estimates_and_time`) to its localization, since there is no such respective feature for names the user enters.

wielinde commented on frontend/app/components/work-packages/wp-single-view/wp-single-view.directive.ts:
```diff
@@ -35,69 +35,104 @@ import {
 
-    return this.singleViewWp.shouldHideField(field, hideEmpty);
+    const hidden = field.visibility === 'hidden';
+    return this.canHideField(field) && (hideEmpty || hidden);
```
@oliverguenther I am not sure what is the issue but I believe this function does not do its job yet. For instance `estimated_time` seems to be always visible although its visibility is set to `default`. So from my understanding, `percentage_done` should only become visible when the user clicks on "Show all attributes". I found that out while comparing what attributes were visible during creation in this version and in the older installations.

wielinde replied:
Another example is `assignee`. That is set to `visible` but does not appear when showing an work package. There more I look at it the more it seems that there is something simply broken.

oliverguenther replied:
Fixed!

wielinde commented on spec/features/types/form_configuration_spec.rb:
```diff
@@ -29,106 +29,223 @@
 
-    it 'the type visibilities are set correctly' do
-      type.reload
+  it 'resets the form properly after changes' do
```
@oliverguenther This spec fails on my machine.

```
 2) form configuration resets the form properly after changes
     Failure/Error: expect(page).to have_selector('.group-name', text: to)
       expected to find css ".group-name" with text "Whatever" but there were no matches. Also found "People", "Estimates and time", "Inactive", which matched the selector but not all filters.
     # ./spec/features/types/form_configuration_spec.rb:115:in `rename_group'
     # ./spec/features/types/form_configuration_spec.rb:138:in `block (2 levels) in <top (required)>'
```

wielinde commented on spec/features/types/form_configuration_spec.rb:
```diff
@@ -29,106 +29,223 @@
+    expect_group('Details')
+  end
+
+  it 'allows modification of the form configuration' do
```
@oliverguenther This spec fails on my machine, too.

```
 form configuration allows modification of the form configuration
     Failure/Error: expect(page).to have_selector('.group-name', text: to)
       expected to find css ".group-name" with text "Whatever" but there were no matches. Also found "People", "Estimates and time", "Inactive", which matched the selector but not all filters.
     # ./spec/features/types/form_configuration_spec.rb:115:in `rename_group'
     # ./spec/features/types/form_configuration_spec.rb:177:in `block (2 levels) in <top (required)>'
```

oliverguenther committed:
Improve validation of group configuration

oliverguenther committed:
Extract visibility into separate class for consistency

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/d2f651cd600fa69ec5ef93246d065715c668f214) (deployed commit: 895105d7714109d2f2ca42ebf9260921ded83d83).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/b091c4ad779153294a2f7b6047131573b3be8fa5) (deployed commit: 8428a2d642d38784f036349c861a5e1d8dfe7ce6).

wielinde reviewed (commented):
Hey Mr Oliver,

I did more of a testing than a code review and I believe I found some heavy bug in the wp-single-view logic. Please find the comments attached.

Wielinde

oliverguenther committed:
Fix hidden fields in new/edit

oliverguenther committed:
Generate default group labels with Setting.default_language

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ece644549ee7dfa578769210f522cbc88e4d7e1b) (deployed commit: 71e4285937461e4d0797af2e396c6d0a1e5fd64a).

oliverguenther committed:
Fix date milestone field

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/be9cede9cab46f0389205bcfaf1447c812b126c2) (deployed commit: ba2c3cadc3dc652781ded7aa50bc8a1f217bd849).

HDinger committed:
Fix stlye issues with type selector

wielinde committed:
Adopt explanatory text to new UI.

wielinde committed:
Extend "Reset to defaults" to reset attribute visibility, too.

wielinde committed:
Prevent validation errors when removing costs plugin for instance

oliverguenther commented:
@wielinde quite some tests are failing now, could you take a look?

wielinde committed:
Fix spec: Problem was that attributes must exist, otherwise they get removed

wielinde committed:
Remove translation logic for group names from form-configuration form

wielinde committed:
Fix spec: attributes must be real

wielinde committed:
Adopt specs to the "self cleaning" feature of the attribute groups

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/83fb0de91942d870bad1ac276185ca4b6f985e38) (deployed commit: 71e830f775d56a9b5cb00202d85d670e2a3f1f5d).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/9c40377b2624b2ceff67d06d48e9262975d6e1a4) (deployed commit: c9512f380ad149633746a6fa19ae2417f3ad9688).

oliverguenther committed:
Avoid enabling new CFs by default

oliverguenther committed:
Allow custom fields to be added through type

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/043df14510d72b4c960350c78367114f86f8ebdd) (deployed commit: e174c8ef3909a5afe1f95a3f4f68b9bc22ecca11).

oliverguenther committed:
Display custom fields in parallel + autoscrolling

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/9f404a09dd093c079bf4b9395d875530a988dd64) (deployed commit: 61d0a976f7571899020b2962a16fbf51949682a1).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/32770eddd6afbd898a82b97184ce776a4952361b) (deployed commit: d26bb2a68d5c23bc544070a1a12d554747731e72).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/5ac1ab8c16b558998c3132612a136e4e87cdb37d) (deployed commit: c31d43cc40ec8a7f6903ecaf5ee13b97b93f0b88).

oliverguenther committed:
Fix specs

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/91a21c8d768eb2260c1856d7eddc45a814653975) (deployed commit: daed93ae33fc572c3b90bf2318e4dc32e92ffa86).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ca417ee518340497efc2175ead063217d57e9707) (deployed commit: 20688eedc9092399b9c654c5bbaa948193c060d6).

HDinger committed:
Change styling of form configuration

wielinde committed:
Some very basic confirm dialog for resetting type form configuration to defaults

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/224ee55a75e6c74ed6d017939e3e2b6f0b853aa9) (deployed commit: 518bd02bc91491605af310928f7c764afd1050fd).

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/211558da6527c9f37f047873e6441af93df9493f) (deployed commit: eba689788f2afac34e0da004994b6e6959c318af).

oliverguenther committed:
Style confirmation dialog and custom fields

oliverguenther committed:
Fix tests

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/1a6f4302c57cab76dd2bcd547299c4b7804546cf) (deployed commit: 790b5b2440d5185cf94fa8659f32d43fd1fe65b9).

oliverguenther committed:
Fix specs after dev merge

crohr commented:
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/6f36a5f336dfc357146320995ee97dc197e9024a) (deployed commit: 0fcb3712328893c05b9c662e5bf87b49e54eade0).
